server/services/v1/billing/event.go

```diff
@@ -132,10 +144,20 @@ func (sb *StorageEventBuilder) Build() *StorageEvent {
 	billingMetric := &StorageEvent{}
 	billingMetric.EventType = EventTypeStorage
 	billingMetric.CustomerId = sb.namespaceId
-	billingMetric.TransactionId = sb.transactionId
-	billingMetric.Timestamp = sb.timestamp
+	// default time
+	if sb.timestamp.IsZero() {
+		sb.timestamp = time.Now()
+	}
+	billingMetric.Timestamp = sb.timestamp.Format(TimeFormat)
 	props := make(map[string]interface{})
 
+	// auto generate transaction id
+	if len(sb.transactionId) != 0 {
```

Comment:
Is this during retries?

Reply:
Just to make sure transaction id is unique at a particular time. Duplicate will be ignored anyways.

Reply:
I was wondering when it is coming from `sb` vs when we need to generate it.